File: 1- Data Structures/1- Basic/3- Queues/queue.cpp
#include <iostream>
#include <queue>

using namespace std;

int main() {
    // Criando uma fila (queue) de inteiros
    queue<int> q;

    // Adicionando elementos à fila
    q.push(10);
    q.push(20);
    q.push(30);

    // Exibindo o elemento no início da fila
    cout << "Primeiro da fila: " << q.front() << endl;  // Output: 10

    // Removendo o elemento do início da fila
    q.pop();
    cout << "Removendo o primeiro, novo primeiro: " << q.front() << endl;  // Output: 20

    // Exibindo o tamanho da fila
    cout << "Tamanho da fila: " << q.size() << endl;  // Output: 2

    // Verificando se a fila está vazia
    if (q.empty()) {
        cout << "A fila está vazia" << endl;
    } else {
        cout << "A fila não está vazia" << endl;  // Output: A fila não está vazia
    }

    return 0;
}

File: 1- Data Structures/1- Basic/3- Queues/queue.js
class Queue {
  constructor() {
      this.items = [];
  }

  // Adicionar um elemento no final da fila
  enqueue(element) {
      this.items.push(element);
  }

  // Remover o elemento do início da fila
  dequeue() {
      if (this.isEmpty()) {
          return "A fila está vazia";
      }
      return this.items.shift();
  }

  // Ver o elemento no início da fila sem remover
  front() {
      if (this.isEmpty()) {
          return "A fila está vazia";
      }
      return this.items[0];
  }

  // Verificar se a fila está vazia
  isEmpty() {
      return this.items.length === 0;
  }

  // Ver o tamanho da fila
  size() {
      return this.items.length;
  }

  // Esvaziar a fila
  clear() {
      this.items = [];
  }
}

// Exemplo de utilização
const queue = new Queue();
queue.enqueue(10);
queue.enqueue(20);
queue.enqueue(30);

console.log("Primeiro da fila:", queue.front());  // Output: 10
console.log("Removendo o primeiro:", queue.dequeue());  // Output: 10
console.log("Novo primeiro da fila:", queue.front());  // Output: 20
console.log("Tamanho da fila:", queue.size());  // Output: 2
